plugins/plugin-bash-like/src/pty/client.ts

```diff
@@ -454,12 +461,13 @@ const webviewChannelFactory: ChannelFactory = async () => {
  * websocket factory for remote/proxy connection
  *
  */
-const getOrCreateChannel = async (
-  cmdline: string,
-  channelFactory: ChannelFactory,
-  tab: Element,
-  terminal: xterm.Terminal
-): Promise<Channel> => {
+const getOrCreateChannel = async (cmdline: string, tab: Tab, terminal: xterm.Terminal): Promise<Channel> => {
+  const channelFactory = inBrowser()
```

Comment:
Maybe we can extract the `channelFactory` to a private constant at `client.ts` file, and avoid the ternary operator to improve the readability.

```
const channelFactory = (): ChannelFactory => {
  if (!inBrowser()) {
    return electronChannelFactory
  }

  if (window['webview-proxy'] !== undefined) {
    return webviewChannelFactory
  }

  return remoteChannelFactory
}
``` 

Or we could use a `ChannelFactory` class with a creational method like the usual implementations of the factory method.

```
const buildRemoteChannel = async () => {
  try {
    const { url, uid, gid }: { url: string; uid: number; gid: number } = await $(
      'bash websocket open',
      undefined,
      undefined,
      {
        rethrowErrors: true
      }
    )
    debug('websocket url', url, uid, gid)
    const WebSocketChannel = (await import('./websocket-channel')).default
    return new WebSocketChannel(url, uid, gid)
  } catch (err) {
    console.error('error opening websocket', err)
    throw err
  }
}

class ChannelFactory {
  static createChannel = async (): Promise<Channel> => {
    if (!inBrowser()) {
      const channel = new InProcessChannel()
      await channel.init()
      return channel
    }

    if (window['webview-proxy'] !== undefined) {
      const channel = new WebViewChannelRendererSide()
      await channel.init()
      return channel
    }

    return buildRemoteChannel()
  }
}
```

This way we only need to invoke `ChannelFactory.createChannel()` at the `getOrCreateChannel` function without including any logic of how it is built.

Reply:
re: static constant. this would be nice, but i haven't been willing yet to guarantee that it is safe to call inBrowser() at the top-level, because this is a capability established at load time.

i personally don't find this code (the current code) so hard to read.
```
  const channelFactory = inBrowser()
    ? window['webview-proxy'] !== undefined
      ? webviewChannelFactory
      : remoteChannelFactory
    : electronChannelFactory
```

but if moving it to a function and avoiding ternaries helps others, i'm happy to comply

Reply:
do we call this method `findChannelFactory`? `getChannelFactory`? it's not `channelFactory` as you have proposed, because then we'd have `const channel = channelFactory()()`, i.e. the function returns a function that produces a channel.

i think i'll keep it as is for now. sometimes, by giving tiny methods names, i think you make the code harder to understand, especially if there is no good name for that method

Reply:
Actually, it is only with the first option, that is implemented with the function wrap to not call the `inBrowser()` on declaration. But the best option is to just call `ChannelFactory.createChannel()` which is implementing the factory method pattern.